### ISSUE TITLE
[Lexer] Number formats supported by lexer differ from number formats supported by literal parser

### ISSUE THREAD

**Issue by kuilpd** [access-softek/llvm-project-dil#4]
As of now DIL lexer has very simple recognition of numbers ([DILLexer.cpp:117](https://github.com/access-softek/llvm-project-dil/blob/b5f75e9c20ec993090ad3d14d4d4fe75c88ad8cb/lldb/source/ValueObject/DILLexer.cpp#L117)) and doesn't recognize any numbers that contain, for example, precision suffix `f` or exponent `e...`, which are recognized and parsed by the literal parser [DILLiteralParsers.cpp:202](https://github.com/access-softek/llvm-project-dil/blob/b5f75e9c20ec993090ad3d14d4d4fe75c88ad8cb/lldb/source/ValueObject/DILLiteralParsers.cpp#L202).
On the other hand, the current literal parser is mostly code from clang, so there is some functionality that we don't need there (such as fractions, half precision, formats `i8` `i16` `i32` etc).
I propose we expand the [numeric_literal](https://github.com/access-softek/llvm-project-dil/blob/b5f75e9c20ec993090ad3d14d4d4fe75c88ad8cb/lldb/docs/dil-expr-lang.ebnf#L152C1-L152C17) in grammar to specify exactly which number formats are supported instead of `clang::tok::numeric_constant`, then add missing formats to lexer and remove unneeded formats from literal parser.

**Comment by asl** [comment]
Tagging @cmtice

The grammar is `numeric_literal = ? clang::tok::numeric_constant ? `. Do we really want all possible clang-recognized numeric literals?

**Comment by cmtice** [comment]
No; that's something that I overlooked correcting. It should be
'numeric_literal = ? dil::TokenKind::numeric_constant ?'.  Thank you for
pointing this out. (The same is true for anywhere else that might mention
'clang::tok::' things).

-- Caroline
***@***.***

**Comment by cmtice** [comment]
BTW...I'm officially OOTO now until Jan. 6, so I may be a bit slow
responding to emails (or not respond at all until I get back).

-- Caroline
***@***.***

**Comment by asl** [comment]
> No; that's something that I overlooked correcting. It should be 'numeric_literal = ? dil::TokenKind::numeric_constant ?'. Thank you for pointing this out. (The same is true for anywhere else that might mention 'clang::tok::' things).

Well, yes. But then `dil::TokenKind::numeric_constant ` is not yet defined. So, the question is how we'd define it. What are the numeric literals we're going to support? Should we just fix the grammar definition (and the lexer) to be aligned with the current implementation of Parser?

**Comment by cmtice** [comment]
I'm leaning towards supporting the numeric literals that are currently tested in the various lldb-eval tests. But I'm open to alternatives...

**Comment by kuilpd** [comment]
@cmtice 
I suggested changes in grammar in #11.
I can start implementing this after holidays, this should solve the majority of failing lldb-eval unit tests I'm launching on DIL.
